### DISCUSSION TITLE
previous definition of the macro `__cmd__greet` here

### DISCUSSION THREAD

**Discussion by gensmusic** [tauri-apps/tauri#4665] [Q&A] [answered]
When I add `pub` before `fn greet` like the folwwing, I got the error
```rust
#[tauri::command]
pub fn greet(name: &str) -> DirEntry {
    format!("Hello, {}!", name);
    DirEntry::new("a")
}
```

error

```
16 | #[tauri::command]
   | ----------------- previous definition of the macro `__cmd__greet` here
17 | pub fn greet(name: &str) -> DirEntry {
   |        ^^^^^ `__cmd__greet` reimported here
   |
   = note: `__cmd__greet` must be defined only once in the macro namespace of this module
help: you can use `as` to change the binding name of the import
   |
17 | pub fn greet as other___cmd__greet(name: &str) -> DirEntry {
   |        ~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Since remove pub the error is gone,  I have to put all the `#[tauri::command]` with `invoke_handler(tauri::generate_handler![greet])` in the same file ?

I want to put `#[tauri::command]` into seperated files and then use them as needed.

**Comment by FabianLars** [comment, marked as answer]
This is only an issue if you use `pub` on commands that are in the main file. If you outsource them in a separate module/file, pub works fine again.

**Reply by gensmusic** [reply]
cannot use `pub` in lib.rs either in my test.

**Reply by FabianLars** [reply]
Yeah because main.rs and lib.rs are, i don't know how to phrase it, interpreted as the same file? Like notice how you don't need to add imports in main.rs for stuff defined in lib.rs? They are somewhat in the same scope idk

That said, lib.rs is a special case, "normal" module files work fine.

**Reply by drmason13** [reply]
Note that the commands must be defined within the binary crate, i.e. in main.rs or a submodule thereof.

Putting commands in a submodule under a lib.rs also doesn't work, in my experience. I assume some aspect of the command macro relies on being used in the same crate as the generate_handler macro is used... which is fair enough!
